LibGfx/JPEG: Support RGB JPEGs

LucasChollet opened the pull request:
That was pretty straightforward :smile:

1. We now decode the App14 segment, where the needed color transformation is stored
2. We apply the color transformation previously read

One small thing to note, we need to go from signed to unsigned for RGB.
The implementation of this conversion shows flaws of the current design by imposing 6 nested loops. Good news, it will be an easy function to test any refactoring!

nico commented:
Huh, I've never seen a single file using this as far as I can remember, and I've looked at quite a few jpegs (their markers, that is, not at the image) in my time. Have you seen this in the wild? Which software can generate these?

(Jpeg has an extremely long tail of extensions that have not adoption at all, which is why I'm asking.)

LucasChollet commented:
I generated this image with `cjpeg -rgb`, and actually I have no idea if it's used somewhere.
The full story behind it is because once I used cjpeg to generate grayscale image and I replaced the -grayscale argument with -rgb because I wanted colors. The result was quite surprising 😅.

I added it because it was a low hanging fruit once support for the App14 segment was implemented.

I'm realizing, are you saying that you have never seen the App14 segment or RGB images ?

nico commented:
I wouldn't add code for stuff we don't see in the wild personally. I've never seen an RGB jpeg (…as far as I know). Maybe let's just close this and then reopen it if we ever come across one?

LucasChollet commented:
Isn't the App14 segment "needed" for CMYK images? (btw, if you have examples of wild CMYK, hmu)
I made this PR as a step forward CMYK support and RGB was only 10 lines of very maintainable code away. So even if it's not used very often, it shouldn't hurt much. Moreover, it means replacing working code with a FIXME, I'm not very keen on the idea.

nico commented on Userland/Libraries/LibGfx/JPEGLoader.cpp:
```diff
@@ -214,6 +214,15 @@ struct Scan {
+    // https://www.itu.int/rec/dologin_pub.asp?lang=e&id=T-REC-T.872-201206-I!!PDF-E&type=items
+    // 6.5.3 - APP14 marker segment for colour encoding
+
+    CmykOrRgb,
```
nit: for things from specs, I usually include the explicit `= 0` (etc) instead of relying on that being there implicitly. but up to you.

very nit: i'd also toss the newline probably

nico commented on Userland/Libraries/LibGfx/JPEGLoader.cpp:
```diff
@@ -787,6 +798,50 @@ static ErrorOr<void> read_icc_profile(SeekableStream& stream, JPEGLoadingContext
+        return Error::from_string_literal("App14 segment too small");
+
+    [[maybe_unused]] auto const version = TRY(stream.read_value<u8>());
+    [[maybe_unused]] u16 const flag0 = static_cast<u16>(TRY(stream.read_value<u8>())) << 8 | TRY(stream.read_value<u8>());
```
`u16 const flag0 = TRY(stream.read_value<BigEndian<u16>>());` instead probably?

nico commented on Userland/Libraries/LibGfx/JPEGLoader.cpp:
```diff
@@ -787,6 +798,50 @@ static ErrorOr<void> read_icc_profile(SeekableStream& stream, JPEGLoadingContext
+
+    [[maybe_unused]] auto const version = TRY(stream.read_value<u8>());
+    [[maybe_unused]] u16 const flag0 = static_cast<u16>(TRY(stream.read_value<u8>())) << 8 | TRY(stream.read_value<u8>());
+    [[maybe_unused]] u16 const flag1 = static_cast<u16>(TRY(stream.read_value<u8>())) << 8 | TRY(stream.read_value<u8>());
```
same

nico replied:
…oh, I suppose that doesn't doesn't work because of unaligned data. Maybe put a comment saying `// Can't use BigEndian<u16> because the data is unaligned`?

LucasChollet replied:
Hmm, I guess that `BigEndian<u16>` should also work. My tests aren't really conclusive as flags are always zeroed in all my images.
So let's yeet those nasty `static_cast` and if there is a world where we care about these flags, we will figure out then

nico reviewed (approved):
Shows what I know! LG :)
